src/main/java/com/urbanairship/datacube/dbharnesses/HBaseDbHarness.java

```diff
@@ -322,43 +391,78 @@ private void flushBatch(Batch<T> batch) throws IOException, InterruptedException
-                final T op = entry.getValue();
-                final byte[] rowKey = ArrayUtils.addAll(uniqueCubeName, address.toWriteKey(idService));
-                final long nanoTimeBeforeWrite = System.nanoTime();
-
-                byte[] dbBytes;
-                switch (commitType) {
-                    case INCREMENT:
-                        long postIncr = increment(rowKey, op);
-                        dbBytes = Bytes.toBytes(postIncr);
-                        break;
-                    case READ_COMBINE_CAS:
-                        dbBytes = readCombineCas(rowKey, op);
-                        break;
-                    case OVERWRITE:
-                        overwrite(rowKey, op);
-                        dbBytes = op.serialize();
-                        break;
-                    default:
-                        throw new RuntimeException("Unsupported commit type " + commitType);
+            if (commitType.equals(CommitType.INCREMENT) && configuration.batchSize > 1) {
+                Iterable<List<Map.Entry<Address, T>>> partitions = Iterables.partition(batchMap.entrySet(), configuration.batchSize);
+                int batchesPerFlush = 0;
+                Map<BoxedByteArray, T> increments = new HashMap<>();
+                Map<BoxedByteArray, Address> backwards = new HashMap<>();
+                for (List<Map.Entry<Address, T>> entries : partitions) {
+                    batchesPerFlush++;
+                    final long nanoTimeBeforeWrite = System.nanoTime();
+                    for (Map.Entry<Address, T> entry : entries) {
+                        final Address address = entry.getKey();
+                        final T op = entry.getValue();
+                        final BoxedByteArray rowKey = new BoxedByteArray(ArrayUtils.addAll(uniqueCubeName, address.toWriteKey(idService)));
+                        increments.put(rowKey, op);
+                        backwards.put(rowKey, address);
+                    }
+
+                    Map<BoxedByteArray, byte[]> successes = increment(increments);
+
+                    for (Map.Entry<BoxedByteArray, byte[]> entry : successes.entrySet()) {
+                        successfulAddresses.add(backwards.get(entry.getKey()));
+                        successfulRows.put(entry.getKey().bytes, entry.getValue());
+                    }
+
+                    long writeDurationNanos = System.nanoTime() - nanoTimeBeforeWrite;
+                    batchWritesTimer.update(writeDurationNanos, TimeUnit.NANOSECONDS);
+                }
+                batchesPerFlushHisto.update(batchesPerFlush);
+
+                if (successfulAddresses.size() < batchMap.size()) {
+                    // the implementation prior to the addition of the batch increment code assumes any failed increment
+                    // operation results in an io exception. This matches that expectation.
+                    throw new IOException("Some writes failed; queueing retry");
```

Comment:
Won't this mean a single failed increment in a batch will result in all the other increments being double applied?

Reply:
Successful increments are removed from the batch before it is retried: https://github.com/urbanairship/datacube/pull/86/files/b8e0046dac6d3ea1aa4779f1f37997e6a2b77383#diff-1494ebe332eeadb4122b501220798d73R476

Reply:
It's probably worth adding some additional metrics on failed puts per flush though

Reply:
Added metric for failure per flush here https://github.com/urbanairship/datacube/pull/86/commits/3b8163b947367e310ea802d5d22e4319324c59fc